example/visualization/tensorboard/mnist.py

```diff
@@ -0,0 +1,160 @@
+    trainer = gluon.Trainer(net.collect_params(), 'sgd',
+                            {'learning_rate': opt.lr, 'momentum': opt.momentum})
+    metric = mx.metric.Accuracy()
+    loss = gluon.loss.SoftmaxCrossEntropyLoss()
+
+    # collect parameter names for logging the gradients of parameters in each epoch
+    params = net.collect_params()
+    param_names = params.keys()
+
+    # define the number of bins for logging histograms
+    num_bins = 1000
+
+    # define a summary writer that logs data and flushes to the file every 5 seconds
+    sw = SummaryWriter(logdir='logs', flush_secs=5)
+
+    for epoch in range(epochs):
+        # reset data iterator and metric at begining of epoch.
+        metric.reset()
+        for i, (data, label) in enumerate(train_data):
+            # Copy data to ctx if necessary
+            data = data.as_in_context(ctx)
+            label = label.as_in_context(ctx)
+            # Start recording computation graph with record() section.
+            # Recorded graphs can then be differentiated with backward.
+            with autograd.record():
+                output = net(data)
+                L = loss(output, label)
+            L.backward()
+
+            # take a gradient step with batch_size equal to data.shape[0]
+            trainer.step(data.shape[0])
+            # update metric at last.
+            metric.update([label], [output])
+
+            if i % opt.log_interval == 0 and i > 0:
+                name, acc = metric.get()
+                print('[Epoch %d Batch %d] Training: %s=%f' % (epoch, i, name, acc))
+
+            # log the first batch of images of each epoch to make sure that we are
+            # training based upon the correct examples and dataset as well as
+            # the dataset is really shuffled in each epoch.
+            if i == 0:
+                sw.add_image(('epoch%d_minibatch%d' % (epoch, i)),
+                             data.reshape((opt.batch_size, 1, 28, 28)), epoch)
+
+        grads = [i.grad() for i in net.collect_params().values()]
+        assert len(grads) == len(param_names)
+        # logging the gradients of parameters for checking convergence
+        for i, name in enumerate(param_names):
+            sw.add_histogram(tag=name, values=grads[i], global_step=epoch, bins=num_bins)
+
+        name, acc = metric.get()
+        print('[Epoch %d] Training: %s=%f' % (epoch, name, acc))
+        # logging training accuracy for visualizing the training accuracy curve
+        sw.add_scalar(tag='train_acc', value=acc, global_step=epoch)
+
+        name, val_acc = test(ctx)
+        # logging the validation accuracy for visualizing the validation accuracy curve
+        print('[Epoch %d] Validation: %s=%f' % (epoch, name, val_acc))
+        sw.add_scalar(tag='valid_acc', value=val_acc, global_step=epoch)
+
+    net.save_params('mnist.params')
+    sw.close()
+
+
+if __name__ == '__main__':
+    if opt.cuda:
+        ctx = mx.gpu(0)
```

Comment:
Would this limit the training to only 1 gpu?

Reply:
No, it would not.